### PR TITLE
Charlesmchen/incremental backup8

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -204,6 +204,7 @@
 		34D1F0BA1F8800D90066283D /* OWSAudioMessageView.m in Sources */ = {isa = PBXBuildFile; fileRef = 34D1F0B91F8800D90066283D /* OWSAudioMessageView.m */; };
 		34D1F0BD1F8D108C0066283D /* AttachmentUploadView.m in Sources */ = {isa = PBXBuildFile; fileRef = 34D1F0BC1F8D108C0066283D /* AttachmentUploadView.m */; };
 		34D1F0C01F8EC1760066283D /* MessageRecipientStatusUtils.swift in Sources */ = {isa = PBXBuildFile; fileRef = 34D1F0BF1F8EC1760066283D /* MessageRecipientStatusUtils.swift */; };
+		34D2CCD220618B3000CB1A14 /* OWSBackupLazyRestoreJob.swift in Sources */ = {isa = PBXBuildFile; fileRef = 34D2CCD120618B2F00CB1A14 /* OWSBackupLazyRestoreJob.swift */; };
```
Comment (charlesmchen) [inline Signal.xcodeproj/project.pbxproj]: This PR implements lazy restore of attachments.

File: Signal/src/AppDelegate.m [hunk 2]
```diff
@@ -418,8 +418,8 @@ - (UIViewController *)loadingRootViewController
     // Every time we change or add a database view in such a way that
     // might cause a delay on launch, we need to bump this constant.
     //
-    // We upgraded YapDatabase in v2.20.0 and need to regenerate all database views.
-    NSString *kLastVersionWithDatabaseViewChange = @"2.20.0";
+    // We added a database view in v2.23.0.
+    NSString *kLastVersionWithDatabaseViewChange = @"2.23.0";
```
Comment (charlesmchen) [inline Signal/src/AppDelegate.m]: Added a view for "attachments that need lazy restore."

File: Signal/src/AppDelegate.m [hunk 3]
```diff
@@ -1180,6 +1180,8 @@ - (void)checkIfAppIsReady
     [self ensureRootViewController];
 
     [OWSBackup.sharedManager setup];
+
+    [OWSBackupLazyRestoreJob run];
```
Comment (charlesmchen) [inline Signal/src/AppDelegate.m]: On app launch we try to resume lazy restore.

File: Signal/src/util/OWSBackup.m [hunk 4]
```diff
@@ -483,6 +484,164 @@ - (void)clearAllCloudKitRecords
         }];
 }
 
+#pragma mark - Lazy Restore
+
+- (NSArray<NSString *> *)attachmentRecordNamesForLazyRestore
```
Comment (charlesmchen) [inline Signal/src/util/OWSBackup.m]: When we "backup export", we need to preserve the records for all attachments that are lazy restoring.

File: Signal/src/util/OWSBackupExportJob.m [hunk 5]
```diff
@@ -346,7 +345,7 @@ - (void)start
         if (self.isComplete) {
             return;
         }
-        [self tryToFetchManifestWithCompletion:^(BOOL tryToFetchManifestSuccess) {
+        [self fetchAllRecordsWithCompletion:^(BOOL tryToFetchManifestSuccess) {
```
Comment (charlesmchen) [inline Signal/src/util/OWSBackupExportJob.m]: Now that we have a local cache of "known backup fragments", we no longer need to download the "last known manifest".

File: Signal/src/util/OWSBackupExportJob.m [hunk 6]
```diff
@@ -295,8 +295,7 @@ @interface OWSBackupExportJob ()
 @property (nonatomic, nullable) OWSBackupExportItem *manifestItem;
 
 // If we are replacing an existing backup, we use some of its contents for continuity.
-@property (nonatomic, nullable) NSDictionary<NSString *, OWSBackupManifestItem *> *lastManifestItemMap;
-@property (nonatomic, nullable) NSSet<NSString *> *lastRecordNames;
```
Comment (charlesmchen) [inline Signal/src/util/OWSBackupExportJob.m]: Superceded by a local cache of "known backup fragments" in the database.

File: Signal/src/util/OWSBackupImportJob.m [hunk 7]
```diff
@@ -248,53 +258,44 @@ - (void)restoreAttachmentFiles
...
+                continue;
+            }
+            TSAttachmentStream *_Nullable attachment =
+                [TSAttachmentStream fetchObjectWithUniqueID:item.attachmentId transaction:transaction];
+            if (!attachment) {
+                DDLogError(@"%@ attachment to restore could not be found.", self.logTag);
+                // Attachment-related errors are recoverable and can be ignored.
+                continue;
+            }
+            [attachment updateWithLazyRestoreFragment:item transaction:transaction];
```
Comment (charlesmchen) [inline Signal/src/util/OWSBackupImportJob.m]: The key statement.  We no longer restore attachments here, instead we _enqueue_ them for lazy restore.

File: Signal/src/util/OWSBackupJob.h [hunk 8]
```diff
@@ -20,31 +24,10 @@ typedef void (^OWSBackupJobCompletion)(NSError *_Nullable error);
 typedef void (^OWSBackupJobManifestSuccess)(OWSBackupManifestContents *manifest);
 typedef void (^OWSBackupJobManifestFailure)(NSError *error);
 
-@interface OWSBackupManifestItem : NSObject
```
Comment (charlesmchen) [inline Signal/src/util/OWSBackupJob.h]: Renamed to `OWSBackupFragment` and moved to SSK.

File: Signal/src/AppDelegate.m (hunk 3 above)
Comment (michaelkirk-signal) [reply Signal/src/AppDelegate.m]: runAsync?

File: Signal/src/util/OWSBackupExportJob.m [hunk 9]
```diff
@@ -818,16 +743,16 @@ - (BOOL)saveNextAttachmentFileToCloudWithCompletion:(OWSBackupJobCompletion)comp
         //   this record's metadata.
         // * That this record does in fact exist in our CloudKit database.
```
Comment (michaelkirk-signal) [inline Signal/src/util/OWSBackupExportJob.m]: Is this comment still up to date now that we don't have lastManifestItemMap?

File: SignalServiceKit/src/Storage/TSDatabaseView.m [hunk 10]
```diff
@@ -341,6 +342,57 @@ + (void)asyncRegisterSecondaryDevicesDatabaseView:(OWSStorage *)storage
...
+            return NSOrderedSame;
+        }
+        if (![object2 isKindOfClass:[TSAttachment class]]) {
+            OWSProdLogAndFail(@"%@ Unexpected entity %@ in collection: %@", self.logTag, [object2 class], collection2);
+            return NSOrderedSame;
+        }
+
+        TSAttachmentStream *attachmentStream1 = (TSAttachmentStream *)object1;
+        TSAttachmentStream *attachmentStream2 = (TSAttachmentStream *)object2;
+        return [attachmentStream2.creationTimestamp compare:attachmentStream1.creationTimestamp];
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Storage/TSDatabaseView.m]: At first glance this looked like a typo, but I think this is intentionally reverse chronological, right? Can you add a comment?

File: SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m [hunk 11]
```diff
@@ -610,6 +613,46 @@ - (CGFloat)audioDurationSeconds
...
+- (void)updateWithLazyRestoreComplete
+{
+    [self.dbReadWriteConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+        [self applyChangeToSelfAndLatestCopy:transaction
+                                 changeBlock:^(TSAttachmentStream *attachment) {
+                                     [attachment setLazyRestoreFragmentId:nil];
+                                 }];
+    }];
+
+    [self ensureThumbnail];
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m]: Is this necessary? 

Won't this happen by virtue of `applyChangeToSelf...` calling  `saveWithTransaction:`?

File: SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m [hunk 12]
```diff
@@ -359,7 +362,7 @@ - (void)ensureThumbnail
     }
 
     if (![[NSFileManager defaultManager] fileExistsAtPath:self.mediaURL.path]) {
-        OWSFail(@"%@ while generating thumbnail, source file doesn't exist: %@", self.logTag, self.mediaURL);
+        DDLogError(@"%@ while generating thumbnail, source file doesn't exist: %@", self.logTag, self.mediaURL);
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m]: Maybe? `OWSAssert(self.lazyRestoreFragmentId)`

File: SignalServiceKit/src/Storage/TSDatabaseView.m [hunk 13]
```diff
@@ -295,30 +300,26 @@ + (YapDatabaseViewSorting *)messagesSorting {
...
-                OWSDevice *device = (OWSDevice *)object;
-                if (![device isPrimaryDevice]) {
-                    return TSSecondaryDevicesGroup;
-                }
+    YapDatabaseViewGrouping *viewGrouping = [YapDatabaseViewGrouping withObjectBlock:^NSString *_Nullable(
+        YapDatabaseReadTransaction *transaction, NSString *collection, NSString *key, id object) {
+        if ([object isKindOfClass:[OWSDevice class]]) {
+            OWSDevice *device = (OWSDevice *)object;
+            if (![device isPrimaryDevice]) {
+                return TSSecondaryDevicesGroup;
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Storage/TSDatabaseView.m]: I'm assuming this is clang-format gone wild.

File: SignalServiceKit/src/Storage/TSDatabaseView.m [hunk 14]
```diff
@@ -295,30 +300,26 @@ + (YapDatabaseViewSorting *)messagesSorting {
...
-            id _Nonnull object2) {
-
+        [YapDatabaseViewSorting withObjectBlock:^NSComparisonResult(YapDatabaseReadTransaction *transaction,
+            NSString *group,
+            NSString *collection1,
+            NSString *key1,
+            id object1,
+            NSString *collection2,
+            NSString *key2,
+            id object2) {
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Storage/TSDatabaseView.m]: I'm assuming this is clang-format gone wild.

File: Signal/src/AppDelegate.m (hunk 3 above)
Comment (charlesmchen-signal) [reply Signal/src/AppDelegate.m]: Done.

File: Signal/src/util/OWSBackupExportJob.m (hunk 9 above)
Comment (charlesmchen-signal) [reply Signal/src/util/OWSBackupExportJob.m]: Fixed.

File: SignalServiceKit/src/Storage/TSDatabaseView.m (hunk 10 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Storage/TSDatabaseView.m]: Added comment.

File: SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m (hunk 11 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m]: Oh great!

File: SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m (hunk 12 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m]: Good idea.

File: SignalServiceKit/src/Storage/TSDatabaseView.m (hunk 13 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Storage/TSDatabaseView.m]: Seems that way.

File: SignalServiceKit/src/Storage/TSDatabaseView.m (hunk 14 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Storage/TSDatabaseView.m]: DatabaseView.m

Comment (charlesmchen-signal) [reply SignalServiceKit/src/Storage/TSDatabaseView.m]: Seems that way.

**Pull request merged**